### PR TITLE
Adding ability do handle MRI metadata to be saved into the BIDS Sidecar

### PR DESCRIPTION
**Pull request by bhatiadheeraj**

### CODE REVIEW

**Comment by anibalsolon** [comment]
@bhatiadheeraj is this still up to review? let us know when it is ready

**Comment by dlevitas** [comment]
@anibalsolon I've alerted Dheeraj to a few lingering issues while reviewing, so we can hold off for now.

**Comment by anibalsolon** [comment]
@bhatiadheeraj @dlevitas I've removed ourselves as reviewers- to be re-added when the PR is good to review. Please let's try to keep reviews through the issue so others can also understand what's happening.

**Comment by dlevitas** [comment]
@bhatiadheeraj, this PR is coming along nicely, there's one lingering issue that I failed to mention during our last meeting.

One of the sequences is `perf/m0scan`, but when clicking on the "Edit Metadata" button, the required fields are as if it's a `perf/asl` sequence. This is probably due to the fact that the bids-specification asl.yaml file also includes information for not just the `asl` suffix, but also `m0scan`, see [here](https://github.com/bids-standard/bids-specification/blob/9f9f64ce43ba92735410507787312550701db267/src/schema/rules/sidecars/asl.yaml#L191). 

Given that we removed the "IntendedFor" metadata field from `perf/m0scan` sequence button, perhaps we can simply exclude this functionality for `perf/m0scan` sequences.

**Comment by dlevitas** [comment]
@bhatiadheeraj disregard my previous message. My recent commit addresses the issue I described.

@anibalsolon feel free to review this PR when you get a chance. I ran through from a user (front-end) perspective, and was able to successfully create and download a BIDS-compliant ASL dataset. I didn't give much consideration to back-end processes, so feel to focus your review on that if you need that's appropriate.

**Comment by dlevitas** [comment]
Something for us to consider, but shouldn't prevent this PR from being merged:

Some metadata fields are optional, but become recommended (or required), depending on the value of other metadata fields. When optional metadata fields become required, they are outlined in red, which alerts the users to this change. If however, optional metadata fields become recommended, the user will be unaware. Additionally, the newly recommended metadata field won't move to the Recommended column. We should eventually address this, but it's not high priority for now, given that this won't trigger a BIDS validation error, nor does it prevent users from entering the information.

**Comment by francopestilli** [comment]
@bhatiadheeraj can you please add a box around each column in the interface. I mean the following. Currently the interface for the metadata has 3 columns: Required, Recommended and Optional. I think it would be helpful to have each column be marked by a very lightly weighted box. Similar to the ezBIDS landing page box see here.
image

**Comment by bhatiadheeraj** [comment]
Ready for review !

File: ui/src/assets/schema/rules/sidecars/asl.yaml [hunk 1]
```diff
@@ -120,7 +141,10 @@ MRIASLPcaslSpecific:
   fields:
     PCASLType:
       level: recommended
-      level_addendum: if `ArterialSpinLabelingType` is `"PCASL"`
+      level_addendum: recommended if `ArterialSpinLabelingType` is `PCASL`
```
Comment (anibalsolon) [inline ui/src/assets/schema/rules/sidecars/asl.yaml:144]: `PCASLType` is already recommended, does the addendum change anything?

File: ui/src/components/modalityForm.vue [hunk 2]
```diff
@@ -201,48 +299,58 @@ export default defineComponent({
...
+                                        // Check if the parsed value is still an empty array
+                                    if(value.length === 0) continue; // skip to the next iteration
+                                }
+                                if(type == 'object') value = JSON.parse(value,key);
+                                json[key] = value;
+                            }
+                        }
+
+                        item.sidecar_json = JSON.stringify(json, null, 2);
+                        // console.log("finalOutput",item.sidecar);
```
Comment (anibalsolon) [inline ui/src/components/modalityForm.vue:340]: need cleanup

File: ui/src/components/modalityForm.vue [hunk 3]
```diff
@@ -201,48 +299,58 @@ export default defineComponent({
...
+                    }
+                });
+            }
+        });
+    },
+    inputType(item) {
+        // console.log("itemDetails",item.details);
```
Comment (anibalsolon) [inline ui/src/components/modalityForm.vue:347]: cleanup

File: ui/src/components/modalityForm.vue [hunk 4]
```diff
@@ -463,6 +575,36 @@ export default defineComponent({
...
+                                    callback(new Error('This field is required based on the condition '+fieldName+' == '+expectedValue1+' or '+expectedValue2));
+                                    return;
+                                }
+                            }
+
+                            // required if required if `ArterialSpinLabelingType` is in [`PCASL`,`CASL`,`PASL`]
+                            // is either PCASL or CASL or any other value without brackets
+                            matches = item.condition.match(/`(\w+)`\s+is\s+in\s+\[(`([\w-]+)`(,\s*`([\w-]+)`)*\s*)+\]/);
+                            if(matches) {
+                                // console.log("matches",matches,this.formData.hasOwnProperty(fieldName), String(this.formData[fieldName]), expectedValues, !value);
```
Comment (anibalsolon) [inline ui/src/components/modalityForm.vue:595]: cleanup

File: ui/src/components/modalityForm.vue [hunk 5]
```diff
@@ -626,7 +793,15 @@ export default defineComponent({
                 return {value: item, label: item};
             });
             return array;
-        }
+        },
+        updateDetailsType(item,newValue) {
+            const extractedData = item.details.anyOf.find((item: any) => item.type == newValue);
+            if(extractedData.items){ 
+                // console.log("upextractedData",extractedData);
```
Comment (anibalsolon) [inline ui/src/components/modalityForm.vue:800]: cleanup

File: ui/src/components/modalityForm.vue [hunk 6]
```diff
@@ -626,7 +793,15 @@ export default defineComponent({
...
+                item.details.items = extractedData.items;
+            }
+            // item.details.type = newValue;
```
Comment (anibalsolon) [inline ui/src/components/modalityForm.vue:803]: cleanup

**Review by anibalsolon — APPROVED** [review]
LGTM

File: ui/src/assets/schema/rules/sidecars/asl.yaml (hunk 1 above)
Comment (bhatiadheeraj) [reply ui/src/assets/schema/rules/sidecars/asl.yaml:144]: @dlevitas

Comment (dlevitas) [reply ui/src/assets/schema/rules/sidecars/asl.yaml:144]: The level_addendum specifies conditions where the metadata might become required (or recommended), based on other metadata values

**Comment by dlevitas** [comment]
@bhatiadheeraj there's no validation to differentiate between number and array-type metadata values. For example, with `perf/asl` data the "PostLabelingDelay" can be either a number or array value, but if the user specifies the type as number and provides an array (or vice versa), this isn't flagged. We should alert users when this occurs.

**Comment by bhatiadheeraj** [comment]
image
image

It should be resolved now ! To see changes please refresh the page. 
Thanks for pointing it out !

**Comment by dlevitas** [comment]
LGTM, I'll merge this PR

**Pull request merged**